service/dataset.py

```diff
@@ -1,27 +1,126 @@
+#%%
 import os
 from flask import abort
 from secrets import get as secrets_get
+from secrets import put as secrets_put
 
 
 DATASETS = {"example": {"type": "local_csv",
                         "local_path": os.path.join(os.path.dirname(__file__), "datasets", "example.csv"),
-                        "key": "csv_details"},
+                        "key": "csv_details",
+                        "budget":3.0},
             "demo_dataverse": {"type": "dataverse",
                                "local_metadata_path": os.path.join(os.path.dirname(__file__),
                                                                    "datasets",
                                                                    "dataverse",
                                                                    "demo_dataverse.yml"),
                                "key": "dataverse_details",
-                               "host": "https://demo.dataverse.org/api/access/datafile/395811"}}
+                               "host": "https://demo.dataverse.org/api/access/datafile/395811",
+                               "budget":3.0}}
 
+#%%
+def query_type_budget(q_type):
+    # Dumb function for now
+    return 1.0
 
 def read(info):
     dataset_name = info["dataset_name"]
 
     if dataset_name not in DATASETS:
         abort(400, "Dataset id {} not found.".format(dataset_name))
+    
     dataset = DATASETS[dataset_name]
+
+    # Validate the secret, extract token
     if dataset["type"] == "dataverse":
         dataset["token"] = secrets_get(name="dataverse:{}".format(info["dataset_name"]))["value"]
 
+    # Check/Decrement the budget before returning dataset
+    # Unclear what behaviour budget decrementing should have
+    # - should it check the type of query, and decrement accordingly?
+    adjusted_budget = dataset["budget"] - query_type_budget(info["query_type"]) 
+    if adjusted_budget >= 0.0:
+        dataset["budget"] = adjusted_budget
+    else:
+        abort(412, "Not enough budget for read. Remaining budget: {}".format(dataset_name))
+
     return {"dataset_type": dataset["type"], dataset["key"]: dataset}
+
+#%%
+def register(dset):
+    dataset_name = dset["dataset_name"]
+
+    if dataset_name in DATASETS:
+        abort(401, "Dataset id {} already exists. Identifies must be unique".format(dataset_name))
+    
+    # Well-formed dset, post checks
+    wf_dset = {}
```

Comment:
It is a service call so it is ok to work off of the dset, no need to replicate it